example/amd_smi_list.cc

```diff
@@ -0,0 +1,169 @@
+          device_count = tmp_device_count;
+          isNIC = false;
+          isGPU = false;
+          isSWITCH = true;
+          CHK_AMDSMI_RET(ret)
+        }
+
+        // Allocate the memory for the device handlers on the socket
+        std::vector<amdsmi_processor_handle> processor_handles(device_count);
+
+        if (isNIC) {
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_NIC, &processor_handles[0], &device_count);
+          CHK_AMDSMI_RET(ret)
+        } else if (isGPU) {
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_AMD_GPU, &processor_handles[0], &device_count);
+          CHK_AMDSMI_RET(ret)
+        } else if (isSWITCH) {
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_SWITCH, &processor_handles[0], &device_count);
+          CHK_AMDSMI_RET(ret)
+        } else {
+          printf("\n Unknown device as part of discovered list. Exiting! \n");
+          ret = amdsmi_shut_down();
+          CHK_AMDSMI_RET(ret)
+          return 0;
+        }
+ 
+        for (uint32_t j = 0; j < device_count; j++) {
+
+            processor_type_t processor_type = {};
+            ret = amdsmi_get_processor_type(processor_handles[j], &processor_type);
+            CHK_AMDSMI_RET(ret)
+
+            amdsmi_bdf_t bdf = {};
+
+            if (isNIC) {
+                ret = amdsmi_get_nic_device_bdf(processor_handles[j], &bdf);
+            }
+            else if (isGPU) {
+                ret = amdsmi_get_gpu_device_bdf(processor_handles[j], &bdf);
+            } 
+            else if (isSWITCH) {
+              ret = amdsmi_get_switch_device_bdf(processor_handles[j], &bdf);
+            }
+            std::string pType = "";
+            if (processor_type == AMDSMI_PROCESSOR_TYPE_AMD_GPU) {
+                pType= "AMD_GPU";
+            }
+            else if(processor_type == AMDSMI_PROCESSOR_TYPE_BRCM_NIC) {
+                pType = "BRCM_NIC";
+            }
+            else if (processor_type == AMDSMI_PROCESSOR_TYPE_BRCM_SWITCH) {
+              pType = "BRCM_SWITCH";
+            }
+            unsigned int uuid_length = AMDSMI_GPU_UUID_SIZE;
+            char uuid[AMDSMI_GPU_UUID_SIZE];
+            if (isNIC) {
+                ret = amdsmi_get_nic_device_uuid(processor_handles[j], &uuid_length, uuid);
+            } 
+            else if (isGPU) {
+                ret = amdsmi_get_gpu_device_uuid(processor_handles[j], &uuid_length, uuid);
+            } 
+            else if (isSWITCH) {
+              ret = amdsmi_get_switch_device_uuid(processor_handles[j], &uuid_length, uuid);
+            }
+            CHK_AMDSMI_RET(ret)
+
+            if(ret != AMDSMI_STATUS_SUCCESS)
+                        std::cout<<"Failed to get bdf"<<"["<<j<<"] , Err["<<ret<<"] "<< std::endl;
+            else
+                printf("\tDevice[%d] \n\ttype[%s] \n\tBDF %04lx:%02x:%02x.%d \n\tUUID:%s\n\n", i,pType.c_str(), 
```

Comment:
Understandable use of `printf` until we move this project to `C++20` standard, where we can use `std::format` and `std::print` in `C++23`.